drivers/tee/optee/optee.c

```diff
@@ -0,0 +1,500 @@
+			tp->attr = TEE_PARAM_ATTR_TYPE_NONE;
+			break;
+		case OPTEE_MSG_ATTR_TYPE_VALUE_INPUT:
+		case OPTEE_MSG_ATTR_TYPE_VALUE_OUTPUT:
+		case OPTEE_MSG_ATTR_TYPE_VALUE_INOUT:
+			tp->attr = TEE_PARAM_ATTR_TYPE_VALUE_INOUT + attr -
+				OPTEE_MSG_ATTR_TYPE_VALUE_INPUT;
+			tp->a = mtp->u.value.a;
+			tp->b = mtp->u.value.b;
+			tp->c = mtp->u.value.c;
+			break;
+		case OPTEE_MSG_ATTR_TYPE_RMEM_INPUT:
+		case OPTEE_MSG_ATTR_TYPE_RMEM_OUTPUT:
+		case OPTEE_MSG_ATTR_TYPE_RMEM_INOUT:
+			tp->attr = TEE_PARAM_ATTR_TYPE_MEMREF_INPUT + attr -
+				OPTEE_MSG_ATTR_TYPE_RMEM_INPUT;
+			tp->b = mtp->u.rmem.size;
+
+			if (!mtp->u.rmem.shm_ref) {
+				tp->a = 0;
+				tp->c = 0;
+				break;
+			}
+
+			tp->a = mtp->u.rmem.offs;
+			tp->c = mtp->u.rmem.shm_ref;
+			break;
+		default:
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+static uint64_t regs_to_u64(uint32_t reg0, uint32_t reg1)
+{
+	return (uint64_t)(((uint64_t)reg0 << 32) | reg1);
+}
+
+static void u64_to_regs(uint64_t val, uint32_t *reg0, uint32_t *reg1)
+{
+	*reg0 = val >> 32;
+	*reg1 = val;
+}
+
+static void handle_rpc_call(const struct device *dev, struct optee_rpc_param *param)
+{
+	struct tee_shm *shm = NULL;
+
+	switch (OPTEE_SMC_RETURN_GET_RPC_FUNC(param->a0)) {
+	case OPTEE_SMC_RPC_FUNC_ALLOC:
+		if (!tee_add_shm(dev, NULL, OPTEE_MSG_NONCONTIG_PAGE_SIZE,
+				 param->a1,
+				 TEE_SHM_ALLOC, &shm)) {
+			u64_to_regs((uint64_t)shm->addr, &param->a1, &param->a2);
+			u64_to_regs((uint64_t)shm, &param->a4, &param->a5);
+		} else {
+			param->a1 = 0;
+			param->a2 = 0;
+			param->a4 = 0;
+			param->a5 = 0;
+		}
+		break;
+	case OPTEE_SMC_RPC_FUNC_FREE:
+		shm = (struct tee_shm *)regs_to_u64(param->a1, param->a2);
+		tee_rm_shm(dev, shm);
+		break;
+	case OPTEE_SMC_RPC_FUNC_FOREIGN_INTR:
+		/* Foreign interrupt was raised */
+		break;
+	case OPTEE_SMC_RPC_FUNC_CMD:
+		/* TODO: Tee supplicatnt function should be called here */
+		break;
+	default:
+		break;
+	}
+
+	param->a0 = OPTEE_SMC_CALL_RETURN_FROM_RPC;
+}
+
+static int optee_call(const struct device *dev, struct optee_msg_arg *arg)
+{
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+	struct optee_rpc_param param = {
+		.a0 = OPTEE_SMC_CALL_WITH_ARG
+	};
+
+	u64_to_regs((uint64_t)arg, &param.a1, &param.a2);
+	while (true) {
+		struct arm_smccc_res res;
+
+		data->smc_call(param.a0, param.a1, param.a2, param.a3,
+			       param.a4, param.a5, param.a6, param.a7, &res);
+
+		if (OPTEE_SMC_RETURN_IS_RPC(res.a0)) {
+			param.a0 = res.a0;
+			param.a1 = res.a1;
+			param.a2 = res.a2;
+			param.a3 = res.a3;
+			handle_rpc_call(dev, &param);
+		} else {
+			return res.a0 == OPTEE_SMC_RETURN_OK ? TEEC_SUCCESS :
+				TEEC_ERROR_BAD_PARAMETERS;
+		}
+	}
+}
+
+static int optee_get_version(const struct device *dev, struct tee_version_info *info)
+{
+	if (!info) {
+		return -EINVAL;
+	}
+
+	info->impl_id = TEE_IMPL_ID_OPTEE;
+	info->impl_caps = TEE_OPTEE_CAP_TZ;
+	info->gen_caps = TEE_GEN_CAP_GP | TEE_GEN_CAP_REG_MEM;
```

Comment:
IIRC, those caps should be read from OP-TEE itself. Registered SHM is an optional build parameter in OP-TEE.

Reply:
Would you please create the same issue on https://github.com/xen-troops/zephyr/? External users do not have access to EPAM Jira.

Also, you need TODO there.

Reply:
Added TODO and created an Issue
https://github.com/xen-troops/zephyr/issues/61